### PR TITLE
KIALI-173 Show nodes that do not yet get traffic

### PR DESCRIPTION
**Pull request by lucasponce**
This PR introduces a non intrusive check to detect services without traffic and add them into the representation of the graph.

The strategy is to fetch the services (and versions) for a namespace and place them in the graph with no traffic.

In this version we didn't enter to "infer" potential sources/destinations from the configuration for several reasons:
- These non-traffic nodes in the graph are potentially ephemeral, as long as they have traffic, they are gone.
- Infer sources and destinations could not be possible from Istio config in a high number of cases, and this could add complexity.
- There are some default scenarios that might cover potential situations well enough.

So, let's describe some scenarios:

a) Namespace with no traffic, no info about sources or destinations (i.e. using the istio-tutorial demo)

image

b) Namespace with partial traffic  (i.e. using the bookinfo namespace, generate traffic on bookinfo app, install the istio-tutorial demo, but leave these services without traffic)

image

On this example, once traffic flows, the graph is updated with proper layout

image

c) Namespace with traffic, and it appears a new version without traffic

image

I think we would cover most of cases of potential situations.

Note, this PR doesn't intend to refactor existing logic, just add an additional check on the existing design.

### CODE REVIEW

**Comment by pilhuhn** [comment]
I need to look into this more. A question though: could we change the inferred edges from a straight line to a dashed one to better differentiate the "we don't know yet" from the "we know, but there was no traffic in the time range" case?

**Comment by lucasponce** [comment]
> A question though: could we change the inferred edges from a straight line to a dashed one to better differentiate the "we don't know yet" from the "we know, but there was no traffic in the time range" case?

I re-used the current graph generation which draws in black when rate is equals 0.
For these nodes we can put a rate equals -1 or a different flag, then it can be draw differently with a dashed edge.

This is a good suggestion IMO.

**Comment by jshaughn** [comment]
There are two things here, really, there are "no-traffic" nodes and "never-traffic" nodes (maybe "unused" or "discovered" is a better term?).   Anyway, "no-traffic" means there has been previous traffic and so a time-series exists and we do know the proper edges.  But for the current duration there has been no traffic.   With "never-traffic" nodes we have never seen traffic, there are no time series, and we don't know what the edges are going to be.  In the example above the never-traffic nodes are linked to "Unknown" because we don't know what to do.  Maybe we should see what it looks like to link them only to a composite node (i.e. create a box around them) with no other edges, just a grouping identifying the services that are discovered but have never been used.

I'm not exactly sure how this would look without seeing it.

**Comment by lucasponce** [comment]
I am working on the "dashed" proposal from @pilhuhn. I think it is more clear, specially with nodes like case c) presented in my first comment, where I have one version with traffic, and other version without it (and no timeseries data in prometheus). On this case, the natural place I would like to see is in the same box as the siblings.

Grouping all "dashed" nodes that comes from the parent "unknown" sounds more like a corner case, I don't know, if it is a good idea, as today, visually a box is to represent some versions of the same service, which can or can not be the case.

Anyway, I think we can work in iterations here, I'm going to continue with the dashed proposal, and we can refine later when we have more experience and perhaps we can introduce UX for these discussions, and I guess at the end we are discussion how visually to represent a different state here.

**Comment by lucasponce** [comment]
> maybe "unused" or "discovered" is a better term?

I am +1 with unused, as yes, the traffic 0 but exists in prometheus and non exist in prometheus is an state it should be covered.

I'll change that in the code as well.

**Comment by lucasponce** [comment]
So, now unused nodes will have a dotted line like this:

image

So, non traffic (rate 0) will be different of unused nodes (rate -1).

Added a counterpart on the UI to propagate the line-style.

File: handlers/graph_unused.go [hunk 1]
```diff
@@ -0,0 +1,144 @@
+package handlers
+
+import (
+	"fmt"
+
+	"github.com/kiali/kiali/config"
+	"github.com/kiali/kiali/graph/tree"
+	"github.com/kiali/kiali/log"
+	"k8s.io/api/apps/v1beta1"
```
Comment (jmazzitelli) [inline handlers/graph_unused.go]: trivial comment - please order the imports the way the rest of the code base is. 

```
import (
  // std imports
  // blank line
  // third party imports
  // blank line
  // our kiali imports
```

File: handlers/graph_unused_test.go [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+
+import (
+	"testing"
+
+	"k8s.io/api/apps/v1beta1"
+	meta_v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/kiali/kiali/config"
+	"github.com/kiali/kiali/graph/tree"
+	"github.com/stretchr/testify/assert"
```
Comment (jmazzitelli) [inline handlers/graph_unused_test.go]: trivial - this third party import is organized incorrectly - put it up in the third-party section

**Comment by jshaughn** [comment]
The dashed lines are an improvement for distinguishing unused services from no-traffic services.  This could certainly serve as a first-pass solution.  But I'm still not sure I agree that, "the natural place I would like to see is in the same box as the siblings".  It depends on what we are trying to bring to the user's attention.  My thought was that 'unused' services represent services that may have an unexpected issue, and as such  we may want to treat them much differently.  We can wait for feedback, it's just a gut feeling and I may be wrong.  It just seems a bit odd to me to link unused services to the 'unknown' service when they may very well not be services that handle external client requests.

**Comment by lucasponce** [comment]
@pilhuhn by default, with no info, I took the unused nodes with a dotted line under the unknown.

Where would you like to see them ?

For me, placing them under unknown is not confusing if the information is clear that those are potential connections, and as soon as we have traffic graph will update.

If there is a different option, I will be happy to apply it.

**Comment by jshaughn** [comment]
"Unknown" basically represents traffic from the internet and it will typically route to the 'ingress' service unless the configuration allows for client requests to go directly to a service.

In Vizceral they actually called this node "internet" but Istio calls it the "unknown" service.

My feeling is still that the "siblings" in this scenario are the unused services, and that that trumps service versions.  It may be worth "boxing" them with a composite node whether we keep the dotted-line arrows from "unknown", or detach.  But again, we can wait for feedback, I'm not saying this should block the PR.

**Comment by lucasponce** [comment]
> "Unknown" basically represents traffic from the internet and it will typically route to the 'ingress' service unless the configuration allows for client requests to go directly to a service.

That is clear. "Unknown" node is the enter point for external traffic. That was my criteria to place them under these ones in absence of info.

Another criteria can be to place them detached, as commented in the rationale first comment of this PR, this information is potentially ephemeral, so using a dotted line to show that those services are available sounds a reasonable decision as well.

**Comment by jmazzitelli** [comment]
The "Unknown" label is, IMO, not what we should show. It does NOT mean "we do not know what the source is for this service" - that could mean "it could be a service in the mesh, or it might not be". But that is not what it means. It just means its coming from an external source not in the service mesh (e.g. the internet).

So that is why I think this is confusing. We see these unused services directly linked to "Unknown" and people who know what Istio means by "unkonwn" are going to think its a service directly exposed to clients on the internet (or at least external to the mesh). But that is not what we mean here when we connect unused services to that unknown node. We really, truly mean the source requests to that service is unknown and might actually be a service within the mesh - we just don't know. But that is different definition than what this istio "unknown" node really means.

**Comment by jmazzitelli** [comment]
Is it possible to have a separate node  to link these unused nodes to?

And I think we need to rename "Unknown" node to something else. "External" or something like that. "Unknown" is confusing, as we can tell by what this PR is trying to do.

**Comment by lucasponce** [comment]
> Is it possible to have a separate node to link these unused nodes to?

I think I am not agree with your suggestion. So I would let the original requestor @pilhuhn to decide.

File: handlers/graph_unused.go [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+		if !ok {
+			log.Warningf("Deployment %s has not a proper app label [%s]", deployment.Name, appLabel)
+			continue
+		}
+		version, ok := deployment.GetObjectMeta().GetLabels()[versionLabel]
+		if !ok {
+			log.Warningf("Deployment %s has not a proper version label [%s]", deployment.Name, versionLabel)
+			continue
+		}
+		staticNode := tree.NewServiceNode(fmt.Sprintf("%s.%s.%s", app, namespaceName, identityDomain), version)
```
Comment (jotak) [inline handlers/graph_unused.go]: The first "%s" is the istio service name, is it assumed to be the same as the app label value? I wonder if we could end with the same problem @xeviknal is hitting for matching deployments and services, which is that we can't rely on such labels.

**Comment by pilhuhn** [comment]
@lucasponce Actually one could in case of a completely new service (that is not part of a group) also just float its node without any edge.

I think we need to revisit the "unknown" / "Ingress" / "Internet" node in a different PR and discussion

**Comment by lucasponce** [comment]
>  Actually one could in case of a completely new service (that is not part of a group) also just float its node without any edge.

So, then your proposal is to leave nodes without edges when it can't find a group ?

**Comment by lucasponce** [comment]
> yes (?) But I am no UxD person :)

I will re-ask to confirm, as what you asked it is the first a) case covered in the first comment of this PR.

So, you want to re-arrange this PR, to show those nodes without connection under "unknown" node.

Is that right ?

**Comment by jmazzitelli** [comment]
I would say just merge this PR. We'll improve as we go.

**Comment by lucasponce** [comment]
> I would say just merge this PR. We'll improve as we go.

Please, don't. Let's clarify ambiguities.

Re-arrange things is not a problem.

**Comment by jshaughn** [comment]
Even though I think there may be a better way to represent the unused services, I'd also recommend a merge. Let people see it in action and comment.  No need to hold the PR, the backend work is fine, presentation changes can be iterative.

If you really don't want to merge, please add then please add screen shots using a composite.

**Comment by lucasponce** [comment]
I commented with @pilhuhn as whatever solution has no cost in terms of implementation.
So, the idea is to show this approach to UX team and collect feedback from them and from users.

Once we have more info we can address this with a different root node (External), no root, or other apporach.

Thanks for the comments.

File: handlers/graph_unused.go (hunk 3 above)
Comment (lucasponce) [reply handlers/graph_unused.go]: As these ones are Istio services, I think we are safe.

**Pull request merged**